src/values/set.ts

```diff
@@ -0,0 +1,181 @@
+import { Client } from "../client";
+import { Query, fql } from "../query-builder";
+import { QuerySuccess, QueryValue } from "../wire-protocol";
+
+/**
+ * A materialize view of a Set.
+ * @see {@link https://fqlx-beta--fauna-docs.netlify.app/fqlx/beta/reference/language/types#set}
+ */
+export class Page<T extends QueryValue> {
+  /** A materialized page of data */
+  readonly data: T[];
+  /**
+   * A pagination cursor, used to obtain additional information in the Set.
+   * If `after` is not provided, then `data` must be present and represents the
+   * last Page in the Set.
+   */
+  readonly after?: string;
+
+  constructor({ data, after }: { data: T[]; after?: string }) {
+    this.data = data;
+    this.after = after;
+  }
+}
+
+/**
+ * A un-materialize Set. Typically received when a materialized Set contains
+ * another set, the EmbeddedSet does not contain any data to avoid potential
+ * issues such as self-reference and infinite recursion
+ * @see {@link https://fqlx-beta--fauna-docs.netlify.app/fqlx/beta/reference/language/types#set}
+ */
+export class EmbeddedSet {
+  /**
+   * A pagination cursor, used to obtain additional information in the Set.
+   */
+  readonly after: string;
+
+  constructor(after: string) {
+    this.after = after;
+  }
+}
+
+/**
+ * A class to provide an iterable API for fetching multiple pages of data, given
+ * a Fauna Set
+ */
+export class SetIterator<T extends QueryValue>
+  implements AsyncGenerator<T[], void, unknown>
+{
+  readonly #generator: AsyncGenerator<T[], void, unknown>;
+
+  constructor(
+    client: Client,
+    initial: Page<T> | EmbeddedSet | (() => Promise<T>)
+  ) {
+    if (initial instanceof Function) {
+      this.#generator = generateFromThunk(client, initial);
+    } else if (initial instanceof Page || initial instanceof EmbeddedSet) {
+      this.#generator = generatePages(client, initial);
+    } else {
+      throw new TypeError(
+        "Expected 'Pageable<QueryValue> | (() => Promise<T>)', but received " +
+          // @ts-expect-error "Property 'constructor' does not exist on type 'never'."
+          // This is okay, we still want to catch weird inputs from JS
+          initial.constructor.name +
+          " " +
+          JSON.stringify(initial)
```

Comment:
this seems like it could blow up.

I don't think you need to include that.

Reply:
In what way? Even variables with literal values have constructors

```javascript
const x = 5
console.log(x.constructor.name) // Number
```

I guess if you pass in an object that has a constructor property this would make a nonsense error. I'd rather take out the `.constructor.name` part and leave the error though.

Types will be enforced by typescript of course, but for JS it will be easy to pass in nonsense. I think this is one of the places that would be good to catch bad input early when otherwise behavior would be undefined.

Reply:
sure i'm not saying remove the validation. I could just imagine a JS object without a contructor field.